data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/index/DynamicIndexManager.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink.opensearch.index;
+
+import com.google.common.collect.ImmutableSet;
+import org.opensearch.OpenSearchException;
+import org.opensearch.action.admin.cluster.settings.ClusterGetSettingsRequest;
+import org.opensearch.action.admin.cluster.settings.ClusterGetSettingsResponse;
+import org.opensearch.client.RequestOptions;
+import org.opensearch.client.RestHighLevelClient;
+import org.opensearch.client.indices.CreateIndexRequest;
+import org.opensearch.client.indices.GetIndexTemplatesRequest;
+import org.opensearch.client.indices.GetIndexTemplatesResponse;
+import org.opensearch.client.indices.IndexTemplateMetadata;
+import org.opensearch.client.indices.IndexTemplatesExistRequest;
+import org.opensearch.client.indices.PutIndexTemplateRequest;
+import org.opensearch.dataprepper.plugins.sink.opensearch.OpenSearchSinkConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.opensearch.dataprepper.plugins.sink.opensearch.index.IndexManager;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.concurrent.TimeUnit;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TimeZone;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class DynamicIndexManager implements IndexManager {
+    private Cache<String, IndexManager> indexManagerCache;
+    final int CACHE_EXPIRE_AFTER_ACCESS_TIME_MINUTES = 30;
+    private final long cacheSizeInKB = 1024;
+    protected RestHighLevelClient restHighLevelClient;
+    protected OpenSearchSinkConfiguration openSearchSinkConfiguration;
+    final IndexType indexType;
+    private final IndexManagerFactory indexManagerFactory;
+
+    public DynamicIndexManager(final IndexType indexType, final RestHighLevelClient restHighLevelClient, final OpenSearchSinkConfiguration openSearchSinkConfiguration){
+        checkNotNull(restHighLevelClient);
+        checkNotNull(openSearchSinkConfiguration);
+	this.indexType = indexType;
+        this.indexManagerFactory = new IndexManagerFactory();
+        this.restHighLevelClient = restHighLevelClient;
+        this.openSearchSinkConfiguration = openSearchSinkConfiguration;
+        CacheBuilder<String, IndexManager> cacheBuilder = CacheBuilder.newBuilder()
+	                .recordStats()
+		        .concurrencyLevel(1)
+		        .maximumWeight(cacheSizeInKB)
+			.expireAfterAccess(CACHE_EXPIRE_AFTER_ACCESS_TIME_MINUTES, TimeUnit.MINUTES)
+	                .weigher((k, v) -> 32);
```

Comment:
Ideally this is the name of the index plus the `IndexManager` which is associated with it.

Conceptually this would be:

```
(k, v) -> k.length() + computeSize(v)
```

But, I think it is unnecessary to compute that size for now. So using `32` here is a good approach. But, please make this `32` a named constant so that it is not some magic number.

Thus this line should become:

```
.weigher((k, v) -> k.length() + APPROXIMATE_INDEX_MANAGER_SIZE);
```